functions/db/deliveryRequests/onCreate.f.js

```diff
@@ -0,0 +1,134 @@
+        };
+
+        // If a specific RA was requested, force push notification
+        if (raInfo.requested) {
+            console.info('A specific RA requested: ' + raInfo.requested);
+            var raRef = accountsRef.child(raInfo.requested);
+
+            return pushNotification(raRef, notification, 'RA');
+        }
+
+        console.info('No specific RA requested, ' + raInfo.pending.length + 
+            ' pending RAs');
+
+        // Compose all promises to get RA snapshots from db
+        var raSnapPromises = [];
+        for (let ra in raInfo.pending) {
+            raSnapPromises.push(getRASnapPromise(accountsRef, raInfo.pending[ra]));
+        }
+
+        // Get all RA snapshot request results and process them
+        return Promise.all(raSnapPromises).then((results) => {
+            var promise = null;
+
+            for (let i in results) {
+                var raSnap = results[i];
+                var available = isAvailable(raSnap.val(), request);
+
+                console.info('RA "' + raSnap.key + '": available=' + available);
+
+                if (available) {
+                    promise = pushNotification(raSnap.ref, notification, 'RA');
+                }
+            }
+
+            // no available RA, send notification back to DA
+            if (!promise) {
+                // update status of the request
+                promise = requestRef.child('status').set(enums.RequestStatus.UNAVAILABLE);
+                console.info('No RA available, updated request status');
+
+                // create pickup unavailable notification
+                notification = {
+                    type: enums.NotificationType.RECURRING_PICKUP_UNAVAILABLE,
+                    content: requestKey
+                };
+                var daRef = dasRef.child(request.donatingAgency);
+                promise = pushNotification(daRef, notification, 'DA');
+            }
+
+            return promise;
+        });
+    });
+
+function pushNotification(accountRef, notification, label) {
+    var promise = accountRef.child('notifications').push(notification);
+    console.info('Notified ' + label + ' "' + accountRef.key + '": '
+        + JSON.stringify(notification));
+    return promise;
+}
+
+// Firebase calls are asynchronous, return promises in order to execute
+// in sequence as we need to.
+function getRASnapPromise(accountsRef, raId) {
+    return accountsRef.child(raId).once('value').then((raSnap) => {
+        return raSnap;
+    });
+}
+
+// Check if the given RA is available for the pickup request
+function isAvailable(ra, request) {
+    // 0-6 for Sun-Sat
+    var requestDay = moment(request.startDate, enums.DateTimeFormat.DATE).day();
+    var requestStart = moment(request.startTime, enums.DateTimeFormat.TIME);
+    var requestEnd = moment(request.endTime, enums.DateTimeFormat.TIME);
+
+    // raDay could be null if no available slots for that day
+    var raDay = ra.availabilities[requestDay];
+
+    if (raDay) {
```

Comment:
How do we handle if a receiving agency has 2 separate openings on a day? I think the availabilities should store a list of `(start, end)` windows for each day, rather than a single `(start, end)` chunk.

Reply:
From the UI design (see RA signup step 3), we only allow RAs to set up a single `(start, end)` chunk for each day. I agree with you that'd be more ideal, but I think this is okay for now.